tensorboard/webapp/widgets/line_chart_v2/lib/scale.ts

```diff
@@ -0,0 +1,196 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {scaleLinear, scaleLog} from 'd3';
+
+import {ScaleType} from './scale_types';
+
+export interface Scale {
+  forward(domain: [number, number], range: [number, number], x: number): number;
+
+  reverse(domain: [number, number], range: [number, number], x: number): number;
+
+  nice(minAndMax: [number, number]): [number, number];
+
+  /**
+   * @param lowAndHigh bounds of the tick
+   * @param sizeGuidance approximate number of the ticks. Depending on the lowAndHigh, it
+   * may return less or more ticks.
+   */
+  ticks(lowAndHigh: [number, number], sizeGuidance: number): number[];
+}
+
+export function createScale(type: ScaleType): Scale {
+  switch (type) {
+    case ScaleType.LINEAR:
+      return new LinearScale();
+    case ScaleType.LOG10:
+      return new Log10Scale();
+    default:
+      throw new RangeError(`ScaleType ${type} not supported.`);
```

Comment:
Let’s bind `const _: never = type;` so that we get a type error here if
we ever add a new scale type. (And then still throw the error, since
TypeScript isn’t sound.)